apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+        if (!_existChallenge(_actionId, challengeId, challenge)) {
+            return true;
+        }
+
+        // If the action was challenged but ruled in favour of the submitter
+        // (dispute rejected by arbitrator) or voided, return true
+        ChallengeState state = challenge.state;
+        return state == ChallengeState.Rejected || state == ChallengeState.Voided;
+    }
+
+    /**
+    * @dev Tell whether an action can be settled
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action can be settled, false otherwise
+    */
+    function _canSettle(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        return _isWaitingChallengeAnswer(_actionId, _challenge);
+    }
+
+    /**
+    * @dev Tell whether an action can be disputed
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action can be disputed, false otherwise
+    */
+    function _canDispute(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        if (!_isWaitingChallengeAnswer(_actionId, _challenge)) {
+            return false;
+        }
+
+        return _challenge.endDate > getTimestamp64();
+    }
+
+    /**
+    * @dev Tell whether an action settlement can be claimed
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action settlement can be claimed, false otherwise
+    */
+    function _canClaimSettlement(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        if (!_isWaitingChallengeAnswer(_actionId, _challenge)) {
+            return false;
+        }
+
+        return getTimestamp64() >= _challenge.endDate;
+    }
+
+    /**
+    * @dev Tell whether an action's dispute can be ruled
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action's dispute can be ruled, false otherwise
+    */
+    function _canRuleDispute(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        return _isDisputed(_actionId, _challenge);
+    }
+
+    /**
+    * @dev Tell whether an action is challenged and if it's waiting to be answered
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action is challenged and it's waiting to be answered, false otherwise
+    */
+    function _isWaitingChallengeAnswer(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        return _actionId == _challenge.actionId && _challenge.state == ChallengeState.Waiting;
```

Comment:
Similar to other comment above, I feel like the `_actionId == _challenge.actionId` is not that useful.

Reply:
For the record, we can probably continue the discussion [here](https://github.com/aragon/aragon-apps/pull/1133#discussion_r439623006)